soroban-simulation/src/snapshot_source.rs

```diff
@@ -115,3 +120,169 @@ impl<T: SnapshotSourceWithArchive> SnapshotSource for AutoRestoringSnapshotSourc
+}
+
+// This is the same as `SimulationSnapshotSource`, but for
+// `SnapshotSourceWithArchive` trait.
+// Note, that we don't implement both traits for `SimulationSnapshotSource` in order
+// to avoid confusion: we never want to accidentally create a `SnapshotSource` with
+// `SnapshotSourceWithArchive` inner snapshot, or vice versa.
+pub(crate) struct SimulationSnapshotSourceWithArchive<'a, T: SnapshotSourceWithArchive> {
+    inner_snapshot: &'a T,
+    entry_updater: RefCell<LedgerEntryUpdater>,
+}
+
+impl<'a> SimulationSnapshotSource<'a> {
+    pub(crate) fn new(snapshot: &'a dyn SnapshotSource) -> Self {
+        Self {
+            inner_snapshot: SnapshotSourceHolder::Ref(snapshot),
+            entry_updater: RefCell::new(Default::default()),
+        }
+    }
+
+    pub(crate) fn new_from_rc(snapshot: Rc<dyn SnapshotSource>) -> Self {
+        Self {
+            inner_snapshot: SnapshotSourceHolder::Rc(snapshot),
+            entry_updater: RefCell::new(Default::default()),
+        }
+    }
+}
+
+impl<'a, T: SnapshotSourceWithArchive> SimulationSnapshotSourceWithArchive<'a, T> {
+    pub(crate) fn new(snapshot: &'a T) -> Self {
+        Self {
+            inner_snapshot: &snapshot,
+            entry_updater: RefCell::new(Default::default()),
+        }
+    }
+}
+
+impl<'a> SnapshotSource for SimulationSnapshotSource<'a> {
+    fn get(&self, key: &Rc<LedgerKey>) -> Result<Option<EntryWithLiveUntil>, HostError> {
+        Ok(self
+            .entry_updater
+            .borrow_mut()
+            .maybe_update_entry(key, self.inner_snapshot.get(key)?))
+    }
+}
+
+impl<'a, T: SnapshotSourceWithArchive> SnapshotSourceWithArchive
+    for SimulationSnapshotSourceWithArchive<'a, T>
+{
+    fn get_including_archived(
+        &self,
+        key: &Rc<LedgerKey>,
+    ) -> Result<Option<EntryWithLiveUntil>, HostError> {
+        Ok(self
+            .entry_updater
+            .borrow_mut()
+            .maybe_update_entry(key, self.inner_snapshot.get_including_archived(key)?))
+    }
+}
+
+fn update_account_entry(account_entry: &mut AccountEntry) {
+    match &mut account_entry.ext {
+        AccountEntryExt::V0 => {
+            let mut ext = AccountEntryExtensionV1 {
+                liabilities: Liabilities {
+                    buying: 0,
+                    selling: 0,
+                },
+                ext: AccountEntryExtensionV1Ext::V0,
+            };
+            fill_account_ext_v2(&mut ext);
+            account_entry.ext = AccountEntryExt::V1(ext);
+        }
+        AccountEntryExt::V1(ext) => {
+            fill_account_ext_v2(ext);
+        }
+    }
+}
+
+fn fill_account_ext_v2(account_ext_v1: &mut AccountEntryExtensionV1) {
+    match &mut account_ext_v1.ext {
+        AccountEntryExtensionV1Ext::V0 => {
+            let mut ext = AccountEntryExtensionV2 {
+                num_sponsored: 0,
+                num_sponsoring: 0,
+                signer_sponsoring_i_ds: Default::default(),
```

Comment:
In core, we resize this vector to the size of the signers vector on the account when the v2 extension is created - https://github.com/stellar/stellar-core/blob/f7a767ac5f6a2f64d9584a1042c57881a57586cf/src/transactions/TransactionUtils.cpp#L83. We should do the same here right?

Reply:
I'm not 100% sure, but it's not hard to support this just in case. I've added the respective logic.